src/rtctools/optimization/seed_mixin.py

```diff
@@ -0,0 +1,126 @@
+"""Module for seeding functionalities.
+"""
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Union
+
+from rtctools.data.timeseries import (
+    check_times_are_equidistant,
+    check_times_are_increasing,
+    fill_nan_in_timeseries,
+    get_timeseries_from_csv,
+    get_timeseries_from_pi,
+)
+from rtctools.optimization.csv_mixin import CSVMixin
+from rtctools.optimization.goal_programming_mixin import GoalProgrammingMixin
+from rtctools.optimization.homotopy_mixin import HomotopyMixin
+from rtctools.optimization.optimization_problem import OptimizationProblem
+from rtctools.optimization.pi_mixin import PIMixin
+from rtctools.optimization.timeseries import Timeseries
+
+DATA_DIR = Path(__file__).parent / "reservoir"
+
+
+@dataclass
+class SeedOptions:
+    """
+    Class describing options for using a seed.
+
+    :cvar seed_timeseries:
+        Path to the file containing the seed.
+        If None, the default seed is used.
+    :cvar is_binary:
+        Indicates if the file is in binary format.
+    :cvar validate_timeseries:
+        Indicates if the timeseries should be validated.
+    :cvar times_are_equidistant:
+        Indicates if the time stamps are equidistant.
+    :cvar fallback:
+        Indicates if the solver should fallback to the default seed
+        if it fails for the given seed.
+    """
+
+    seed_timeseries: Union[Path, None] = None
+    is_binary: bool = False
+    validate_timeseries: bool = False
+    times_are_equidistant: bool = False
+    fallback: bool = False
+
+
+class SeedMixin(OptimizationProblem):
```

Comment:
I'm still not clear on the purpose of this new mixin.

Currently, if a time series is specified for an optimization variable in `timeseries_import.csv`, its values are taken as seed values.

All this mixin seems to do is provide a new pathway to specify seed values on top and above of the existing pathway. It would therefore seem to contradict the "Zen of Python" which states that "There should be one-- and preferably only one --obvious way to do it." (https://peps.python.org/pep-0020/).

Reply:
> I'm still not clear on the purpose of this new mixin.
> 
> Currently, if a time series is specified for an optimization variable in `timeseries_import.csv`, its values are taken as seed values.
> 
> All this mixin seems to do is provide a new pathway to specify seed values on top and above of the existing pathway. It would therefore seem to contradict the "Zen of Python" which states that "There should be one-- and preferably only one --obvious way to do it." (https://peps.python.org/pep-0020/).

From what I understand, timeseries values for optimization variables are currently converted into constraints, while values for states are converted into seeds. We have cases where we would like to mix these two, for example to have a constraint for the first time step of a state and a seed for all the other timesteps, or to use constraints on a control variable for certain time steps and a seed for others. This seems currently not possible.

It is also not obvious currently if a timeseries value will be used as a seed or a constraint. Using a separate seed file makes it more explicit what will be used for a seed. It also allows to easily switch between seeds without having to copy all other timeseries. The seed values are also different from constraints / fixed input values, since they do not define the problem, but are more a solver option.

We agree that ideally there should be one obvious way to do it. Perhaps we can decide that for the future, the seed should always be defined in a separate file, but to avoid breaking changes, we still allow seeds in the timeseries_import file.

Reply:
@jbaayen: before I address the other comments, I think it's convenient if we first resolve this issue. Please let me know if you agree with the points above.

Reply:
@SGeeversAtVortech what makes you think that provided timeseries values for optimization variables are turned into constraints?

Reply:
AFAIK the only case where a time series is used as a constraint is when it is labeled 'variable_name + "_Min"' or 'variable_name + "_Max"', in which case the values are used to construct lower respectively upper bounds for the named variable.

Reply:
It is also added as a constraint if the variable is a controlled input

Reply:
I did some code digging and I think the confusion comes from src\rtctools\optimization\collocated_integrated_optimization_problem.py in the part that starts with `# Initial conditions specified in history timeseries`:

```python
            # Initial conditions specified in history timeseries
            history = self.history(ensemble_member)
            for variable in itertools.chain(
                self.differentiated_states, self.algebraic_states, self.controls
            ):
                try:
                    history_timeseries = history[variable]
                except KeyError:
                    pass
                else:
                    interpolation_method = self.interpolation_method(variable)
                    val = self.interpolate(
                        t0,
                        history_timeseries.times,
                        history_timeseries.values,
                        np.nan,
                        np.nan,
                        interpolation_method,
                    )
                    val /= self.variable_nominal(variable)

                    if not np.isnan(val):
                        idx = self.__indices_as_lists[ensemble_member][variable][0]

                        if val < lbx[idx] or val > ubx[idx]:
                            logger.warning(
                                "Initial value {} for variable '{}' outside bounds.".format(
                                    val, variable
                                )
                            )

                        lbx[idx] = ubx[idx] = val
```

It seems that values from before and including the initial time are converted into constraints and later values are used as a seed.

Reply:
In any case, the feature that we try to implement here is to select a seed by filename and automatically try another seed when the optimization fails. Currently, the user would still need to carefully copy+paste historic data of the seeded timeseries + all data of the other timeseries to a separate file and rerun the optimization from scratch, which is rather cumbersome and prone to errors. So although it is technically possible to switch seed, I would not say there is an obvious way to do this right now.

Reply:
This is almost correct: the value from the time series at t0 will be used as initial condition. But there are no constraints created for t < t0 and for t > t0.

The appropriate way to switch seed during a run is to override the `seed()` method. An example of a `seed()` override implementation can be found in `GoalProgrammingMixin`.

Reply:
The user could of course implement his own `seed()` and `optimize()` function as done here in `SeedMixin`, but this implementation is not trivial and has a few pitfalls. The point of this feature, and most other features in rtc-tools, is to make life easier for the user. Probably you can solve any optimization problem with just the `OptimizationProblem` class, but the other mixins are there to provide useful tools and reduce the amount of work on the users side.

Reply:
Yes, but we don't provide dedicated mixins to override each and every overridable function individually. If we did that, we'd have 50 or so different mixins.

The design philosophy has been that mixins add new significant and broadly reusable pieces of functionality that are not available in the core classes.

May I ask you to again try to explain in a few paragraphs what exactly the value-add of the SeedMixin is, considering explicitly that it is currently already possible to provide seed data in a user-friendly manner (e.g., via timeseries_import.csv)?

Reply:
The way I see it, the SeedMixin takes care of running/rerunning an optimization problem for different seeds. It enables the following.

- Enable to select and switch between different seeds, without having to redefine the problem input (I would not consider seeds problem input, since they do not affect the problem definition).
- Enable automatically rerunning if a seed failed and enable adding logic to select the next seed.

For the first property, we would now need to copy all other timeseries in timeseries_import.csv and be careful with the initial values just to use another seed. For the second property, the user would now need to manually check why a run failed (because of the seed or not) and then rerun for different input if necessary.

I agree that we should not add a new mixin when the same can be achieved with a few single functions and the current mixins.

Reply:
I think you are saying that the core value-add is the support of multiple seeds per run, which introduces robustness for cases where the first seed does not result in a convergent run. This makes sense.

How about renaming the mixin to 'MultiSeedMixin' to stress this point?

Reply:
A 'MultiSeedMixin' sounds reasonable.